File: tests/py/include/standard_messages.py
#!/usr/bin/env python3
"""
Standard test message definitions (Python).
Provides get_message(index) function for test messages.

This file matches the C++ standard_messages.hpp structure.
"""

import sys
import os
from typing import List, Union

# Add generated directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'generated', 'py'))

from struct_frame.generated.serialization_test import (
    SerializationTestSerializationTestMessage,
    SerializationTestBasicTypesMessage,
    SerializationTestUnionTestMessage,
    SerializationTestComprehensiveArrayMessage,
    SerializationTestSensor,
    SerializationTestStatus,
    SerializationTestVariableSingleArray,
    SerializationTestMessage,
    SerializationTestMsgSeverity,
    get_message_info,
)

# Type alias for message union (like C++ MessageVariant)
MessageType = Union[
    SerializationTestSerializationTestMessage,
    SerializationTestBasicTypesMessage,
    SerializationTestUnionTestMessage,
    SerializationTestVariableSingleArray,
    SerializationTestMessage,
]

# Message count
MESSAGE_COUNT = 17


# ============================================================================
# Helper functions to create messages (like C++ create_* functions)
# ============================================================================

def create_serialization_test(magic: int, string: str, flt: float, bl: bool, arr: List[int]) -> SerializationTestSerializationTestMessage:
    return SerializationTestSerializationTestMessage(
        magic_number=magic,
        test_string=string.encode('utf-8'),
        test_float=flt,
        test_bool=bl,
        test_array=arr
    )


def create_basic_types(si: int, mi: int, ri: int, li: int, su: int, mu: int, ru: int, lu: int,
                       sp: float, dp: float, fl: bool, dev: str, desc: str) -> SerializationTestBasicTypesMessage:
    return SerializationTestBasicTypesMessage(
        small_int=si,
        medium_int=mi,
        regular_int=ri,
        large_int=li,
        small_uint=su,
        medium_uint=mu,
        regular_uint=ru,
        large_uint=lu,
        single_precision=sp,
        double_precision=dp,
        flag=fl,
        device_id=dev.encode('utf-8'),
        description=desc.encode('utf-8')
    )


def create_union_with_array() -> SerializationTestUnionTestMessage:
    arr = SerializationTestComprehensiveArrayMessage(
        fixed_ints=[10, 20, 30],
        fixed_floats=[1.5, 2.5],
        fixed_bools=[True, False, True, False],
        bounded_uints=[100, 200],
        bounded_doubles=[3.14159],
        fixed_strings=[b'Hello', b'World'],
        bounded_strings=[b'Test'],
        fixed_statuses=[SerializationTestStatus.ACTIVE.value, SerializationTestStatus.ERROR.value],
        bounded_statuses=[SerializationTestStatus.INACTIVE.value],
        fixed_sensors=[SerializationTestSensor(id=1, value=25.5, status=SerializationTestStatus.ACTIVE.value, name=b'TempSensor')],
        bounded_sensors=[]
    )
    return SerializationTestUnionTestMessage(
        payload={'array_payload': arr},
        payload_which='array_payload',
        payload_discriminator=SerializationTestComprehensiveArrayMessage.MSG_ID
    )


def create_union_with_test() -> SerializationTestUnionTestMessage:
    test = SerializationTestSerializationTestMessage(
        magic_number=0x12345678,
        test_string=b'Union test message',
        test_float=99.99,
        test_bool=True,
        test_array=[1, 2, 3, 4, 5]
    )
    return SerializationTestUnionTestMessage(
        payload={'test_payload': test},
        payload_which='test_payload',
        payload_discriminator=SerializationTestSerializationTestMessage.MSG_ID
    )


def create_variable_single_array_empty() -> SerializationTestVariableSingleArray:
    return SerializationTestVariableSingleArray(message_id=0x00000001, payload=[], checksum=0x0001)


def create_variable_single_array_single() -> SerializationTestVariableSingleArray:
    return SerializationTestVariableSingleArray(message_id=0x00000002, payload=[42], checksum=0x0002)


def create_variable_single_array_third() -> SerializationTestVariableSingleArray:
    return SerializationTestVariableSingleArray(message_id=0x00000003, payload=list(range(67)), checksum=0x0003)


def create_variable_single_array_almost() -> SerializationTestVariableSingleArray:
    return SerializationTestVariableSingleArray(message_id=0x00000004, payload=list(range(199)), checksum=0x0004)


def create_variable_single_array_full() -> SerializationTestVariableSingleArray:
    return SerializationTestVariableSingleArray(message_id=0x00000005, payload=list(range(200)), checksum=0x0005)


def create_message_test() -> SerializationTestMessage:
    return SerializationTestMessage(
        severity=SerializationTestMsgSeverity.SEV_MSG.value,
        module=b"test",
        msg=b"A really good"
    )


# ============================================================================
# get_message(index) - unified interface matching C++ MessageProvider pattern
# ============================================================================

def get_message(index: int) -> MessageType:
    """Get message by index."""
    if index == 0:
        return create_serialization_test(0xDEADBEEF, "Cross-platform test!", 3.14159, True, [100, 200, 300])
    elif index == 1:
        return create_serialization_test(0, "", 0.0, False, [])
    elif index == 2:
        return create_serialization_test(0xFFFFFFFF, "Maximum length test string for coverage!", 999999.9, True,
                                         [2147483647, -2147483648, 0, 1, -1])
    elif index == 3:
        return create_serialization_test(0xAAAAAAAA, "Negative test", -273.15, False, [-100, -200, -300, -400])
    elif index == 4:
        return create_serialization_test(1234567890, "Special: !@#$%^&*()", 2.71828, True, [0, 1, 1, 2, 3])
    elif index == 5:
        return create_basic_types(42, 1000, 123456, 9876543210, 200, 50000, 4000000000, 9223372036854775807,
                                  3.14159, 2.718281828459045, True, "DEVICE-001", "Basic test values")
    elif index == 6:
        return create_basic_types(0, 0, 0, 0, 0, 0, 0, 0, 0.0, 0.0, False, "", "")
    elif index == 7:
        return create_basic_types(-128, -32768, -2147483648, -9223372036854775807, 255, 65535, 4294967295, 9223372036854775807,
                                  -273.15, -9999.999999, False, "NEG-TEST", "Negative and max values")
    elif index == 8:
        return create_union_with_array()
    elif index == 9:
        return create_union_with_test()
    elif index == 10:
        return create_basic_types(-128, -32768, -2147483648, -9223372036854775807, 255, 65535, 4294967295, 9223372036854775807,
                                  -273.15, -9999.999999, False, "NEG-TEST", "Negative and max values")
    elif index == 11:
        return create_variable_single_array_empty()
    elif index == 12:
        return create_variable_single_array_single()
    elif index == 13:
        return create_variable_single_array_third()
    elif index == 14:
        return create_variable_single_array_almost()
    elif index == 15:
        return create_variable_single_array_full()
    else:  # index == 16
        return create_message_test()

File: tests/csharp/include/StandardMessages.cs
/**
 * Standard test message definitions (C#).
 * Provides GetMessage(index) function for test messages.
 *
 * This file matches the C++ standard_messages.hpp structure.
 */

using System;
using System.Text;
using StructFrame;
using StructFrame.SerializationTest;

namespace StructFrameTests
{
    /// <summary>
    /// Message provider struct for standard test messages.
    /// Provides MESSAGE_COUNT and GetMessage(index) function matching C++ pattern.
    /// </summary>
    public static class StandardMessages
    {
        // Message count
        public const int MESSAGE_COUNT = 17;

        // ============================================================================
        // Helper functions to create messages (like C++ create_* functions)
        // ============================================================================

        private static SerializationTestSerializationTestMessage CreateSerializationTest(
            uint magic, string str, float flt, bool bl, int[] arr)
        {
            var msg = new SerializationTestSerializationTestMessage();
            msg.MagicNumber = magic;
            msg.TestFloat = flt;
            msg.TestBool = bl;

            var stringBytes = Encoding.UTF8.GetBytes(str);
            msg.TestStringLength = (byte)Math.Min(stringBytes.Length, 64);
            msg.TestStringData = new byte[64];
            Array.Copy(stringBytes, msg.TestStringData, msg.TestStringLength);

            msg.TestArrayCount = (byte)Math.Min(arr.Length, 5);
            msg.TestArrayData = new int[5];
            for (int i = 0; i < msg.TestArrayCount; i++)
                msg.TestArrayData[i] = arr[i];

            return msg;
        }

        private static SerializationTestBasicTypesMessage CreateBasicTypes(
            sbyte si, short mi, int ri, long li,
            byte su, ushort mu, uint ru, ulong lu,
            float sp, double dp, bool fl, string dev, string desc)
        {
            var msg = new SerializationTestBasicTypesMessage();
            msg.SmallInt = si;
            msg.MediumInt = mi;
            msg.RegularInt = ri;
            msg.LargeInt = li;
            msg.SmallUint = su;
            msg.MediumUint = mu;
            msg.RegularUint = ru;
            msg.LargeUint = lu;
            msg.SinglePrecision = sp;
            msg.DoublePrecision = dp;
            msg.Flag = fl;

            var deviceIdBytes = Encoding.UTF8.GetBytes(dev);
            msg.DeviceId = new byte[32];
            Array.Copy(deviceIdBytes, msg.DeviceId, Math.Min(deviceIdBytes.Length, 32));

            var descBytes = Encoding.UTF8.GetBytes(desc);
            msg.DescriptionLength = (byte)Math.Min(descBytes.Length, 128);
            msg.DescriptionData = new byte[128];
            Array.Copy(descBytes, msg.DescriptionData, msg.DescriptionLength);

            return msg;
        }

        private static SerializationTestUnionTestMessage CreateUnionWithArray()
        {
            var msg = new SerializationTestUnionTestMessage();
            msg.PayloadDiscriminator = SerializationTestComprehensiveArrayMessage.MsgId;

            var arr = new SerializationTestComprehensiveArrayMessage();
            
            arr.FixedInts = new int[] { 10, 20, 30 };
            arr.FixedFloats = new float[] { 1.5f, 2.5f };
            arr.FixedBools = new bool[] { true, false, true, false };
            
            arr.BoundedUintsCount = 2;
            arr.BoundedUintsData = new ushort[] { 100, 200, 0 };
            
            arr.BoundedDoublesCount = 1;
            arr.BoundedDoublesData = new double[] { 3.14159, 0 };
            
            // Fixed strings: 2 strings, 8 chars each
            arr.FixedStrings = new byte[16];
            var hello = Encoding.UTF8.GetBytes("Hello");
            var world = Encoding.UTF8.GetBytes("World");
            Array.Copy(hello, 0, arr.FixedStrings, 0, Math.Min(hello.Length, 8));
            Array.Copy(world, 0, arr.FixedStrings, 8, Math.Min(world.Length, 8));
            
            // Bounded strings: 1 string
            arr.BoundedStringsCount = 1;
            arr.BoundedStringsData = new byte[24];
            var test = Encoding.UTF8.GetBytes("Test");
            Array.Copy(test, 0, arr.BoundedStringsData, 0, Math.Min(test.Length, 12));
            
            arr.FixedStatuses = new byte[] { (byte)SerializationTestStatus.ACTIVE, (byte)SerializationTestStatus.ERROR };
            
            arr.BoundedStatusesCount = 1;
            arr.BoundedStatusesData = new byte[] { (byte)SerializationTestStatus.INACTIVE, 0 };
            
            arr.FixedSensors = new SerializationTestSensor[1];
            arr.FixedSensors[0] = new SerializationTestSensor
            {
                Id = 1,
                Value = 25.5f,
                Status = SerializationTestStatus.ACTIVE,
                Name = new byte[16]
            };
            var sensorName = Encoding.UTF8.GetBytes("TempSensor");
            Array.Copy(sensorName, arr.FixedSensors[0].Name, Math.Min(sensorName.Length, 16));
            
            arr.BoundedSensorsCount = 0;
            arr.BoundedSensorsData = new SerializationTestSensor[1];

            msg.ArrayPayload = arr;
            return msg;
        }

        private static SerializationTestUnionTestMessage CreateUnionWithTest()
        {
            var msg = new SerializationTestUnionTestMessage();
            msg.PayloadDiscriminator = SerializationTestSerializationTestMessage.MsgId;

            msg.TestPayload = CreateSerializationTest(
                0x12345678,
                "Union test message",
                99.99f,
                true,
                new int[] { 1, 2, 3, 4, 5 }
            );

            return msg;
        }

        private static SerializationTestVariableSingleArray CreateVariableSingleArrayEmpty()
        {
            var msg = new SerializationTestVariableSingleArray();
            msg.MessageId = 0x00000001;
            msg.PayloadCount = 0;
            msg.PayloadData = new byte[200];
            msg.Checksum = 0x0001;
            return msg;
        }

        private static SerializationTestVariableSingleArray CreateVariableSingleArraySingle()
        {
            var msg = new SerializationTestVariableSingleArray();
            msg.MessageId = 0x00000002;
            msg.PayloadCount = 1;
            msg.PayloadData = new byte[200];
            msg.PayloadData[0] = 42;
            msg.Checksum = 0x0002;
            return msg;
        }

        private static SerializationTestVariableSingleArray CreateVariableSingleArrayThird()
        {
            var msg = new SerializationTestVariableSingleArray();
            msg.MessageId = 0x00000003;
            msg.PayloadCount = 67;
            msg.PayloadData = new byte[200];
            for (int i = 0; i < 67; i++)
                msg.PayloadData[i] = (byte)i;
            msg.Checksum = 0x0003;
            return msg;
        }

        private static SerializationTestVariableSingleArray CreateVariableSingleArrayAlmost()
        {
            var msg = new SerializationTestVariableSingleArray();
            msg.MessageId = 0x00000004;
            msg.PayloadCount = 199;
            msg.PayloadData = new byte[200];
            for (int i = 0; i < 199; i++)
                msg.PayloadData[i] = (byte)i;
            msg.Checksum = 0x0004;
            return msg;
        }

        private static SerializationTestVariableSingleArray CreateVariableSingleArrayFull()
        {
            var msg = new SerializationTestVariableSingleArray();
            msg.MessageId = 0x00000005;
            msg.PayloadCount = 200;
            msg.PayloadData = new byte[200];
            for (int i = 0; i < 200; i++)
                msg.PayloadData[i] = (byte)i;
            msg.Checksum = 0x0005;
            return msg;
        }

        private static SerializationTestMessage CreateMessageTest()
        {
            var msg = new SerializationTestMessage();
            msg.Severity = SerializationTestMsgSeverity.SEV_MSG;
            var module = "test";
            msg.ModuleLength = (byte)module.Length;
            msg.ModuleData = Encoding.UTF8.GetBytes(module);
            var msgText = "A really good";
            msg.MsgLength = (byte)msgText.Length;
            msg.MsgData = Encoding.UTF8.GetBytes(msgText);
            return msg;
        }

        // ============================================================================
        // GetMessage(index) - unified interface matching C++ MessageProvider pattern
        // ============================================================================

        public static IStructFrameMessage GetMessage(int index)
        {
            switch (index)
            {
                case 0: return CreateSerializationTest(0xDEADBEEF, "Cross-platform test!", 3.14159f, true, new int[] { 100, 200, 300 });
                case 1: return CreateSerializationTest(0, "", 0.0f, false, new int[] { });
                case 2: return CreateSerializationTest(0xFFFFFFFF, "Maximum length test string for coverage!", 999999.9f, true, new int[] { 2147483647, -2147483648, 0, 1, -1 });
                case 3: return CreateSerializationTest(0xAAAAAAAA, "Negative test", -273.15f, false, new int[] { -100, -200, -300, -400 });
                case 4: return CreateSerializationTest(1234567890, "Special: !@#$%^&*()", 2.71828f, true, new int[] { 0, 1, 1, 2, 3 });
                case 5: return CreateBasicTypes(42, 1000, 123456, 9876543210L, 200, 50000, 4000000000, 9223372036854775807UL, 3.14159f, 2.718281828459045, true, "DEVICE-001", "Basic test values");
                case 6: return CreateBasicTypes(0, 0, 0, 0L, 0, 0, 0, 0UL, 0.0f, 0.0, false, "", "");
                case 7: return CreateBasicTypes(-128, -32768, -2147483648, -9223372036854775807L, 255, 65535, 4294967295, 9223372036854775807UL, -273.15f, -9999.999999, false, "NEG-TEST", "Negative and max values");
                case 8: return CreateUnionWithArray();
                case 9: return CreateUnionWithTest();
                case 10: return CreateBasicTypes(-128, -32768, -2147483648, -9223372036854775807L, 255, 65535, 4294967295, 9223372036854775807UL, -273.15f, -9999.999999, false, "NEG-TEST", "Negative and max values");
                case 11: return CreateVariableSingleArrayEmpty();
                case 12: return CreateVariableSingleArraySingle();
                case 13: return CreateVariableSingleArrayThird();
                case 14: return CreateVariableSingleArrayAlmost();
                case 15: return CreateVariableSingleArrayFull();
                default: return CreateMessageTest();
            }
        }
    }
}
